---
title: "Pleiotropy simulations"
output: html_notebook
---

```{r}
library(simulateGP)
library(tidyverse)
library(lavaan)
library(RadialMR)
library(pbapply)
library(TwoSampleMR)
library(pROC)
```


# Basic simulation setup

This function does the following

1. For each population make a genetic map for all instruments - independent allele frequencies per SNP and population
2. Generate SNP-exposure effects for pop 1 based on h2 and S
3. Transfer the same SNP-exposure effects to pop 2 (the h2 might be different because the allele frequencies are independent)
4. Create SNP-outcome effects for pop1 and pop2 given the causal effects `biv1` and `biv2` and the horizontal pleiotropic effects `pl1` and `pl2`
5. Generate genetic summary data for exposure and outcome traits for pop1 and pop2 given sample sizes `nid1x`, `nid2x`, `nid1y` and `nid2y`
6. Estimate causal relationships in each population using
  - IVW for each pop
  - Radial MR for each pop
  - `UnweightedSEMa` = SEM where biv1=biv2
  - `UnweightedSEMb` = SEM where biv1 and biv2 can vary
  - `RadialSEMa` = SEM in a radialMR framework, e.g. allowing it to be weighted, where biv1=biv2
  - `RadialSEMb` = SEM in a radialMR framework, e.g. allowing it to be weighted, where biv1 and biv2 can vary

```{r}
sim1 <- function(nsnp, nid1x, nid2x, nid1y, nid2y, h2, S, biv1, biv2, pl1, pl2, absx, gx_same=TRUE, instrument_threshold=1, winnerscurse_adjustment=FALSE)
{
	# Create map of variants
	map1 <- tibble(snp=1:nsnp, af=runif(nsnp, 0.05, 0.95))
	map2 <- tibble(snp=1:nsnp, af=runif(nsnp, 0.05, 0.95))

	# Create SNP-exposure effects
	x1 <- generate_gwas_params(map1, h2=h2, Pi=1, S=S)
	if(absx)
	{
		x1$beta <- abs(x1$beta)
	}
	x2 <- map2
	if(gx_same)
	{
    x2$beta <- x1$beta
	} else {
	  x2 <- generate_gwas_params(map2, h2=h2, Pi=1, S=S)
	}

	# Create SNP-outcome effects
	y1 <- x1
	y1$beta <- y1$beta * biv1 + pl1
	y2 <- x2
	y2$beta <- y2$beta * biv2 + pl2

	x1s <- generate_gwas_ss(x1, nid=nid1x)
	x2s <- generate_gwas_ss(x2, nid=nid2x)
	y1s <- generate_gwas_ss(y1, nid=nid1y)
	y2s <- generate_gwas_ss(y2, nid=nid2y)

	sig <- x1s$pval < instrument_threshold | x2s$pval < instrument_threshold

	x1s <- x1s[sig,]
	x2s <- x2s[sig,]
	y1s <- y1s[sig,]
	y2s <- y2s[sig,]

	x1 <- x1[sig,]
	x2 <- x2[sig,]
	y1 <- y1[sig,]
	y2 <- y2[sig,]
	
	if(winnerscurse_adjustment)
	{
	  x1s$bhat <- rnorm(sum(sig), x1$beta, x1s$se)
	  x2s$bhat <- rnorm(sum(sig), x2$beta, x2s$se)
	}
	
	dat <- tibble(
		y1 = y1s$bhat,
		y2 = y2s$bhat,
		x1 = x1s$bhat,
		x2 = x2s$bhat,
		yse1 = y1s$se,
		yse2 = y2s$se,
		xse1 = x1s$se,
		xse2 = x2s$se,
		r1 = y1/x1,
		r2 = y2/x2,
		w1 = sqrt(x1^2 / yse1^2),
		w2 = sqrt(x2^2 / yse2^2),
		o1 = r1 * w1,
		o2 = r2 * w2
	)

	out <- list()
	out$ivw1 <- TwoSampleMR::mr_ivw(dat$x1, dat$y1, dat$xse1, dat$yse1) %>%
	  {tibble(model="IVW", pop=1, bivhat=.$b, se=.$se, pval=.$pval)}
	out$ivw2 <- TwoSampleMR::mr_ivw(dat$x2, dat$y2, dat$xse2, dat$yse2) %>%
	  {tibble(model="IVW", pop=2, bivhat=.$b, se=.$se, pval=.$pval)}
	out$rm1 <- summary(lm(o1 ~ -1 + w1, data=dat)) %>%
	  {tibble(model="RadialIVW", pop=1, bivhat=.$coef[1,1], se=.$coef[1,2], pval=.$coef[1,4])}
	out$rm2 <- summary(lm(o2 ~ -1 + w2, data=dat)) %>%
	  {tibble(model="RadialIVW", pop=2, bivhat=.$coef[1,1], se=.$coef[1,2], pval=.$coef[1,4])}

	
	runsem <- function(model, data, modname)
	{
	  mod <- sem(model, data=data)
	  invisible(capture.output(mod <- summary(mod, fit.measures=TRUE)))
	  tibble(
	    model=modname,
	    pop=1:2,
	    bivhat=mod$PE$est[1:2],
	    se=mod$PE$se[1:2],
	    pval=mod$PE$pvalue[1:2],
	    aic=mod$FIT['aic']
	  )
	}

	out$semA <- runsem('
	y1 ~ biv*x1
	y2 ~ biv*x2
	', dat, "UnweightedSEMa")

	out$semB <- runsem('
	y1 ~ biv_1*x1
	y2 ~ biv_2*x2
	', dat, "UnweightedSEMb")
	
	out$modC <- runsem('
	o1 ~ biv*w1
	o2 ~ biv*w2
	', dat, "RadialSEMa")
	
	out$modD <- runsem('
	o1 ~ biv_1*w1
	o2 ~ biv_2*w2
	', dat, "RadialSEMb")

  compare_est <- function(y, s)
  {
    k <- length(y)
    w <- 1/s^2
    sum.w <- sum(w)
    mu.hat <- sum(y * w)/sum.w
    Q <- sum(w * (y - mu.hat)^2)
    return(Q)
  }

	res <- out %>% bind_rows()
	
	m1 <- lapply(c("IVW", "RadialIVW"), function(x)
	{
	  subset(res, model==x) %>%
	    {
	      compare_est(.$bivhat[1:2], .$se[1:2])
	    } %>%
	    {tibble(model=x, stat=.)}
	}) %>% bind_rows()

	m2 <- tibble(model="RadialSEM", stat=subset(res, model=="RadialSEMa" & pop==1)$aic - subset(res, model=="RadialSEMb" & pop==1)$aic)
	m3 <- tibble(model="UnweightedSEM", stat=subset(res, model=="UnweightedSEMa" & pop==1)$aic - subset(res, model=="UnweightedSEMb" & pop==1)$aic)
		
	popdiff <- bind_rows(m1, m2, m3)
		
	return(list(dat=dat, res=res, popdiff=popdiff))
}
```


Example simulation

```{r}
out <- sim1(
  nsnp=100,
  nid1x=10000,
  nid2x=10000,
  nid1y=10000,
  nid2y=10000,
  h2=0.3,
  S=0,
  biv1=0.2,
  biv2=0.2,
  pl1=0,
  pl2=0,
  absx=FALSE
)
```

What does the data look like:

```{r}
out$dat %>%
  {bind_rows(
    tibble(pop=1, x=.$x1, y=.$y1),
    tibble(pop=2, x=.$x2, y=.$y2)
  )} %>%
  ggplot(., aes(x=x, y=y)) +
    geom_point() +
    geom_smooth(method=lm) +
    facet_grid(. ~ pop)
```
i.e. no directional pleiotropy (slope passes through origin). Estimates:

```{r}
out$res
```

Run a few sims

```{r}
o <- pblapply(1:100, function(i)
{
  sim1(
    nsnp=100,
    nid1x=10000,
    nid2x=10000,
    nid1y=10000,
    nid2y=10000,
    h2=0.3,
    S=0,
    biv1=0.2,
    biv2=0.2,
    pl1=0,
    pl2=0,
    absx=FALSE
  ) %>%
    {.$res} %>%
    mutate(sim=i)
}) %>% bind_rows()
```

```{r}
o %>% 
  group_by(model, pop) %>%
  summarise(bivhat=mean(bivhat), se=mean(se))

o %>%
  ggplot(., aes(x=model, y=bivhat)) +
  geom_hline(yintercept=0.2) +
  geom_boxplot(aes(fill=as.factor(pop)))
```
Summary:

- Everything works as expected (returning estimates ~ 0.2)
- UnweightedSEMa has better precision because it is jointly estimating biv from two independent datasets - power has improved
- Need to note that unweighted MR / SEM will be unstable if allele frequencies are very small

ToDo: 

- Create IVW estimate combining pop1 and pop2 to compare against UnweightedSEMa

## Try different gx


```{r}
o2 <- pblapply(1:100, function(i)
{
  sim1(
    nsnp=100,
    nid1x=10000,
    nid2x=10000,
    nid1y=10000,
    nid2y=10000,
    h2=0.3,
    S=0,
    biv1=0.2,
    biv2=0.2,
    pl1=0,
    pl2=0,
    absx=FALSE,
    gx_same=FALSE
  ) %>%
    {.$res} %>%
    mutate(sim=i)
}) %>% bind_rows()
```

```{r}
o2 %>% 
  group_by(model, pop) %>%
  summarise(bivhat=mean(bivhat), se=mean(se))

o2 %>%
  ggplot(., aes(x=model, y=bivhat)) +
  geom_hline(yintercept=0.2) +
  geom_boxplot(aes(fill=as.factor(pop)))
```
## Instrument significance

```{r}
o31 <- pblapply(1:100, function(i)
{
  sim1(
    nsnp=100,
    nid1x=10000,
    nid2x=10000,
    nid1y=10000,
    nid2y=10000,
    h2=0.3,
    S=0,
    biv1=0.2,
    biv2=0.2,
    pl1=0,
    pl2=0,
    absx=FALSE,
    gx_same=FALSE,
    instrument_threshold = 5e-8,
    winnerscurse_adjustment = TRUE
  ) %>%
    {.$res} %>%
    mutate(sim=i, wc=TRUE)
}) %>% bind_rows()

o32 <- pblapply(1:100, function(i)
{
  sim1(
    nsnp=100,
    nid1x=10000,
    nid2x=10000,
    nid1y=10000,
    nid2y=10000,
    h2=0.3,
    S=0,
    biv1=0.2,
    biv2=0.2,
    pl1=0,
    pl2=0,
    absx=FALSE,
    gx_same=FALSE,
    instrument_threshold = 5e-8,
    winnerscurse_adjustment = FALSE
  ) %>%
    {.$res} %>%
    mutate(sim=i, wc=FALSE)
}) %>% bind_rows()

o3 <- rbind(o31, o32)
o3 %>% 
  group_by(model, pop, wc) %>%
  summarise(bivhat=mean(bivhat), se=mean(se))

o3 %>%
  ggplot(., aes(x=model, y=bivhat)) +
  geom_hline(yintercept=0.2) +
  geom_boxplot(aes(fill=as.factor(pop))) +
  facet_grid(. ~ wc)
```
- Winner's curse seems to be introducing a slight downward bias.



## Introducing pleiotropy

```{r}
out <- sim1(
  nsnp=100,
  nid1x=10000,
  nid2x=10000,
  nid1y=10000,
  nid2y=10000,
  h2=0.3,
  S=0,
  biv1=0.2,
  biv2=0.2,
  pl1=0.2,
  pl2=0.2,
  absx=TRUE
)
```


```{r}
out$dat %>%
  {bind_rows(
    tibble(pop=1, x=.$x1, y=.$y1),
    tibble(pop=2, x=.$x2, y=.$y2)
  )} %>%
  ggplot(., aes(x=x, y=y)) +
    geom_point() +
    geom_smooth(method=lm) +
    facet_grid(. ~ pop)
```

Intercept is now through 0.2 instead of 0 - directional pleiotropy

```{r}
out$res
```
Somewhat unexpected that unweighted SEM is actually unbiased?!

Run a few more times

```{r}
op <- pblapply(1:100, function(i)
{
  sim1(
    nsnp=100,
    nid1x=10000,
    nid2x=10000,
    nid1y=10000,
    nid2y=10000,
    h2=0.3,
    S=0,
    biv1=0.2,
    biv2=0.2,
    pl1=0.2,
    pl2=0.2,
    absx=TRUE
  ) %>%
    {.$res} %>%
    mutate(sim=i)
}) %>% bind_rows()
```



```{r}
op %>% 
  group_by(model, pop) %>%
  summarise(bivhat=mean(bivhat), se=mean(se))

op %>%
  ggplot(., aes(x=model, y=bivhat)) +
  geom_hline(yintercept=0.2) +
  geom_boxplot(aes(fill=as.factor(pop)))
```

Summary

- The unweightedSEM method handles directional pleiotropy automatically


## Population-specific effects

Simulate where the effect is either null in both populations or $b_{pop} \sim N(0, 0.1)$. For IVW and RadialIVW the comparison is made using Cochran's Q test for heterogeneity, to determine if the population 1 and population 2 estimates are non-overlapping. For the SEM methods it is tested using the difference in AIC measures of fit.

```{r}
od <- pblapply(1:500, function(i)
{
  biv1=sample(c(0, rnorm(1, sd=0.01)), 1)
  biv2=sample(c(0, rnorm(1, sd=0.01)), 1)
  sim1(
    nsnp=100,
    nid1x=10000,
    nid2x=10000,
    nid1y=10000,
    nid2y=10000,
    h2=0.3,
    S=0,
    biv1=biv1,
    biv2=biv2,
    pl1=0,
    pl2=0,
    absx=FALSE
  ) %>%
    {.$popdiff} %>%
    mutate(sim=i, biv1=biv1, biv2=biv2)
}) %>% bind_rows()
```

How well does the difference statistic for each method discriminate between cases where the causal effects are the same or different?

```{r}
od %>%
  mutate(same=as.numeric(biv1==biv2)) %>%
  group_by(model) %>%
  summarise(o=as.numeric(auc(response=same, predictor=stat)))
```
Comparison across methods

```{r}
od %>%
  subset(., biv1 != biv2) %>%
  dplyr::select(model, stat, sim) %>%
  pivot_wider(id_cols=c(sim, model), names_from=model, values_from=stat) %>%
  select(-sim) %>%
  pairs()
```
Comparison by effect size difference

```{r}
od %>%
  mutate(diff=abs(biv1-biv2)) %>%
  ggplot(., aes(x=diff, y=stat)) +
    geom_point() +
    facet_grid(model ~ .)
```

Use p-value of 0.05 for Q stat to declare significantly different. What should be used for AIC?

```{r}
aic_thresh <- od %>%
  subset(., biv1==biv2 & model == "UnweightedSEM") %>%
  summarise(t=quantile(stat, 0.95)) %>%
  {.$t}
aic_thresh
```

```{r}
od <- od %>%
  mutate(same=biv1==biv2, sig=case_when(
    model %in% c("IVW", "RadialIVW") ~ pchisq(stat, df=1, low=FALSE) < 0.05,
    model %in% c("UnweightedSEM", "RadialSEM") ~ stat > aic_thresh
  ))


od %>%
  filter(same) %>%
  group_by(model) %>%
  summarise(fdr=sum(sig)/n())

od %>%
  filter(!same) %>%
  group_by(model) %>%
  summarise(power=sum(sig)/n())
```

```{r}
od %>%
  filter(biv1 != biv2) %>%
  ggplot(., aes(x=abs(biv1-biv2), y=as.numeric(sig))) +
  geom_point(aes(colour=model)) +
  geom_smooth(aes(colour=model), se=FALSE) +
  scale_colour_brewer(type="qual") +
  labs(x="Difference in effect on pop1 and pop2", y="Difference detected")
```

Summary

- The UnweightedSEM method, IVW and RadialIVW methods seem to perform comparably

Todo:

- Power of test statistic when effect sizes are different
- Power to detect when SNP-exposure effects are different
- LD-aware simulations: how well can we discriminate between distinct causal variants vs false negatives due to power?


---

# Older stuff


## Further analysis of population specific estimates

```{r}
param <- expand.grid(
  
)
```


```{r}
effects1 <- effects2 <- choose_effects(200, 0.4)
effects1[1:50] <- 0
effects2[51:100] <- 0

# create genotypes
g1 <- make_geno(10000, 200, 0.5)
g2 <- make_geno(10000, 200, 0.5)

# create confounder
u1 <- rnorm(10000)
u2 <- rnorm(10000)

# create x
x1 <- make_phen(c(0.1, effects1), cbind(u1, g1))
x2 <- make_phen(c(0.1, effects2), cbind(u2, g2))

plei <- choose_effects(200, 0.4)

# create y
y1 <- make_phen(c(0.1, 0.2, plei), cbind(u1, x1, g1))
y2 <- make_phen(c(0.1, 0.2, plei), cbind(u2, x2, g2))

# Summary data for pop1 
ss1 <- get_effs(x1, y1, g1)

# Summary data for pop2
ss2 <- get_effs(x2, y2, g2)

mres1 <- mr(ss1, method="mr_ivw")
mres2 <- mr(ss2, method="mr_ivw")


plot(beta.outcome ~ beta.exposure, ss1)
abline(a=0, b=1)
abline(a=0, b=mres1$b, col="red")

# Organise the data
dat <- tibble(
	y1 = ss1$beta.outcome,
	y2 = ss2$beta.outcome,
	x1 = ss1$beta.exposure,
	x2 = ss2$beta.exposure
)

dat$xb <- rowMeans(cbind(dat$x1, dat$x2))
dat$xb[1:100] <- 0
dat$x1[1:50] <- 0
dat$x1[101:200] <- 0
dat$x2[51:100] <- 0
dat$x2[101:200] <- 0

model <- '
y1 ~ b_1*x1 + b_1*xb + p_1*x2
y2 ~ p_2*x1 + b_2*xb + b_2*x2
x1 ~~ 0 * x2
x1 ~~ 0 * xb
x2 ~~ 0 * xb
'
cor(dat)
```
```{r}
mod <- sem(model, data=dat)
summary(mod)
```


```{r}
semPaths(mod, edge.label.cex = 0.7, edge.label.position=0.7)
```

```{r}
nsnp <- 200
map <- tibble(
	snp=1:200,
	af=0.5
) %>%
	mutate(beta = 
		case_when(
			snp > 
	))


xp <- generate_gwas_params(map, 0.4)

x1p <- xp %>%
	mutate(beta = 
		case_when(
			snp <= 50 ~ 0,
			TRUE ~ beta
		)
	)

y1p <- x1p %>%
	mutate(beta = beta * 0.2 + 0.1)

x2p <- xp %>%
	mutate(beta = 
		case_when(
			snp > 50 & snp <= 100 ~ 0,
			TRUE ~ beta
		)
	)

y2p <- x2p %>%
	mutate(beta = beta * 0.2 + 0.1)

yp <- xp %>%
	mutate(beta = beta * 0.2)

summary(lm(yp$beta ~ -1 + xp$beta))
summary(lm(y1p$beta ~ -1 + x1p$beta))
summary(lm(y2p$beta ~ -1 + x2p$beta))

y1 <- generate_gwas_ss(y1p, 1000000)
y2 <- generate_gwas_ss(y2p, 1000000)
x1 <- generate_gwas_ss(x1p, 1000000)
x2 <- generate_gwas_ss(x2p, 1000000)

dat <- tibble(
	y1 = y1$bhat,
	y2 = y2$bhat,
	x1 = x1$bhat,
	x2 = x2$bhat
)

dat$xb <- rowMeans(cbind(dat$x1, dat$x2))
dat$xb[1:100] <- 0
dat$x1[1:50] <- 0
dat$x1[101:200] <- 0
dat$x2[51:100] <- 0
dat$x2[101:200] <- 0

summary(lm(cbind(y1, y2) ~ -1 + x1 + x2 + xb, dat))


summary(lm(y1$bhat ~ -1 + x1$bhat))


model <- '
# linear model
y1 ~ b11*x1 + b11*xb + p1*x2
y2 ~ p2*x1 + b22*xb + b22*x2
'

mod <- sem(model, data=dat, meanstructure=FALSE)
summary(mod)


model <- '
# linear model
y1 ~ b11*x1 + b11*xb + p1*x2
y2 ~ p2*x1 + b22*xb + b22*x2
'

mod <- sem(model, data=dat, meanstructure=FALSE)
summary(mod)


dat2 <- tibble(
	w1 = abs(x1$bhat) / y1$se,
	w2 = abs(x2$bhat) / y2$se,
	Y1 = y1$bhat / x1$bhat * w1,
	Y2 = y2$bhat / x2$bhat * w2,
)

summary(lm(Y1 ~ -1 + w1, dat2))
summary(lm(y1$bhat ~ -1 + x1$bhat, weight=1/y1$se^2))





model <- '
y1 ~ xb + x1
'
summary(sem(model, data=dat))



up <- map %>%
	generate_gwas_params(., 0.1)

xp <- up %>%
	mutate(beta=beta * 0.2)

yp <- up %>%
	mutate(beta=beta * 0.4 + beta * 0.3)

x1p <- xp %>%
	mutate(beta = 
		case_when(
			snp <= 50 ~ 0,
			TRUE ~ beta
		)
	)

y1p <- x1p %>%
	mutate(beta = case_when(
		beta == 0 ~ beta * 0.4,
		TRUE ~ beta * 0.4 + beta * 0.3
	))

x2p <- xp %>%
	mutate(beta = 
		case_when(
			snp > 50 & snp <= 100 ~ 0,
			TRUE ~ beta
		)
	)

y2p <- x2p %>%
	mutate(beta = case_when(
		beta == 0 ~ beta * 0.4,
		TRUE ~ beta * 0.4 + beta * 0.3
	))

plot(y1p$beta ~ x1p$beta)

summary(lm(yp$beta ~ xp$beta))

summary(lm(y2p$beta ~ x1p$beta, subset=x1p$beta == 0))



####


# create genotypes
g1 <- make_geno(10000, 200, 0.5)
g2 <- make_geno(10000, 200, 0.5)

# create confounder
plei1 <- plei2 <- choose_effects(200, 0.4)
effects1[1:50] <- 0
effects2[51:100] <- 0
u1 <- make_phen(plei, g1)
u2 <- make_phen(plei, g2)

x1 <- make_phen(0.6, u1)
x2 <- make_phen(0.6, u2)

y1 <- make_phen(c(0.6, -0.4), cbind(u1, x1))
y2 <- make_phen(c(0.6, -0.4), cbind(u2, x2))


# Summary data for pop1 
ss1 <- get_effs(x1, y1, g1)

# Summary data for pop2
ss2 <- get_effs(x2, y2, g2)

(mres1 <- mr(ss1, method="mr_ivw"))
(mres2 <- mr(ss2, method="mr_ivw"))


plot(beta.outcome ~ beta.exposure, ss1)
abline(a=0, b=1)
abline(a=0, b=mres1$b, col="red")

# Organise the data
dat <- tibble(
	y1 = ss1$beta.outcome,
	y2 = ss2$beta.outcome,
	x1 = ss1$beta.exposure,
	x2 = ss2$beta.exposure
)

dat$xb <- rowMeans(cbind(dat$x1, dat$x2))
dat$xb[1:100] <- 0
dat$x1[1:50] <- 0
dat$x1[101:200] <- 0
dat$x2[51:100] <- 0
dat$x2[101:200] <- 0


model <- '
y1 ~ b_1*x1 + b_1*xb + p_1*x2
y2 ~ p_2*x1 + b_2*xb + b_2*x2
'

mod <- sem(model, data=dat)
summary(mod)


########

# try again


# create genotypes
g1 <- make_geno(10000, 50, 0.5)
g2 <- make_geno(10000, 50, 0.5)
gb <- make_geno(10000, 100, 0.5)

eff1 <- choose_effects(50, 0.2)
eff2 <- choose_effects(50, 0.2)
effb <- choose_effects(100, 0.2)
x1 <- make_phen(eff1, g1)
x2 <- make_phen(eff2, g2)
xb <- make_phen(effb, gb)

plei1 <- choose_effects(50, 0.1)
plei2 <- choose_effects(50, 0.1)
y1 <- make_phen(c(eff1, eff2, effb, 0.2, 0.2), cbind(g1, g2, gb, x1, xb))
y2 <- make_phen(c(eff1, eff2, effb, 0.2, 0.2), cbind(g1, g2, gb, x2, xb))




# Summary data for pop1 
ss1 <- get_effs(x1, y1, g1)

# Summary data for pop2
ss2 <- get_effs(x2, y2, g2) %>% mutate(SNP = SNP + 50)

ssb1 <- get_effs(xb, y1, gb) %>% mutate(SNP = SNP + 100)
ssb2 <- get_effs(xb, y2, gb) %>% mutate(SNP = SNP + 100)


mres1 <- mr(ss1, method="mr_ivw")
mres2 <- mr(ss2, method="mr_ivw")


plot(beta.outcome ~ beta.exposure, ss1)
abline(a=0, b=1)
abline(a=0, b=mres1$b, col="red")

# Organise the data
dat <- tibble(
	y1 = ss1$beta.outcome,
	y2 = ss2$beta.outcome,
	x1 = ss1$beta.exposure,
	x2 = ss2$beta.exposure
)

dat$xb <- rowMeans(cbind(dat$x1, dat$x2))
dat$xb[1:100] <- 0
dat$x1[1:50] <- 0
dat$x1[101:200] <- 0
dat$x2[51:100] <- 0
dat$x2[101:200] <- 0





model <- '
y1 ~ b_1*x1 + b_1*xb + p_1*x2
y2 ~ p_2*x1 + b_2*xb + b_2*x2
x1 ~~ 0 * x2
x1 ~~ 0 * xb
x2 ~~ 0 * xb
'

mod <- sem(model, data=dat)
summary(mod)














effx <- choose_effects(50, 0.2)
effp <- (rnorm(50, sd=sd(effx)/2) + effx) %>% scale() %>% {. * sd(effx)}
cor(effx, effp)

g <- make_geno(10000, 50, 0.5)
colnames(g) <- paste0("g", 1:50)
p <- make_phen(effp, g)
x <- make_phen(effx, g)
y <- make_phen(c(0.2,-0.2), cbind(x, p))

d <- get_effs(x,y,g)
(r <- mr(d, method="mr_ivw"))
mr_scatter_plot(r, d)

library(glue)
mod <- glue('
	x ~ {paste(paste0("g", 1:50), collapse="+")}
	y ~ b * x
')
dat <- data.frame(g, x=x, y=y)
m <- sem(mod, data=dat)
summary(m)

summary(lm(y ~ x))














n <- 100
d <- tibble(
	g1 = rnorm(n),
	g2 = rnorm(n),
	gb = rnorm(n),
	x1 = rnorm(n),
	x2 = rnorm(n),
	xb = rnorm(n),
	y1 = rnorm(n),
	y2 = rnorm(n)
)

m <- '
y1 ~ b1 * x1 + b1 * xb
y2 ~ b2 * x2 + b2 * xb
x1 ~ g1
xb ~ gb
x2 ~ g2
g1 ~~ 0 * gb
g1 ~~ 0 * g2
g2 ~~ 0 * gb
'

mod <- sem(m, data=d)
semPlot::semPaths(mod, layout="tree2")
semPlot::semPaths(mod, layout="circle")


lay <- tidySEM::get_layout(
"g1", "", "gb", "", "g2",
"x1", "", "xb", "", "x2",
"", "y1", "", "y2", "", rows=3
)

tidySEM::graph_sem(mod, layout=lay)

```





```