---
title: "Toy simulations for using SEM to jointly model ancestry-specific effects and pleiotropy"
output: html_notebook
---

```{r}
suppressPackageStartupMessages(suppressWarnings({
library(tidyverse)
library(ieugwasr)
library(simulateGP)
library(lavaan)
library(TwoSampleMR)
library(semPlot)
library(tidySEM)
library(MASS)
}))
```

## Simulations

### Different causal effects in the two populations

Simple model

```{r}
sim1 <- function(nsnp, hsq1, hsq2, rg, nid1, nid2, maf1, maf2, biv1, biv2, inst_pval1, inst_pval2)
{
  # Generate effects for exposures in pop1 and pop2
  s1 <- sqrt(hsq1 / nsnp)
  s2 <- sqrt(hsq2 / nsnp)
  sigma <- matrix(c(1, rg, rg, 1), 2, 2)
  eff <- mvrnorm(nsnp, c(0,0), sigma)
  
  gx1 <- eff[,1] * s1
  gx2 <- eff[,2] * s2

  # create genotypes
  g1 <- make_geno(nid1, nsnp, maf1)
  g2 <- make_geno(nid2, nsnp, maf2)
  
  # create confounder
  u1 <- rnorm(nid1)
  u2 <- rnorm(nid2)
  
  # create x
  x1 <- make_phen(c(0.1, gx1), cbind(u1, g1))
  x2 <- make_phen(c(0.1, gx2), cbind(u2, g2))

  # create y
  y1 <- make_phen(c(0.1, biv1), cbind(u1, x1))
  y2 <- make_phen(c(0.1, biv2), cbind(u2, x2))

  # Summary data for pop1 
  ss1 <- get_effs(x1, y1, g1)
  
  # Summary data for pop2
  ss2 <- get_effs(x2, y2, g2)

  # Organise the data
  dat <- tibble(
  	y1 = ss1$beta.outcome,
  	y2 = ss2$beta.outcome,
  	x1 = ss1$beta.exposure,
  	x2 = ss2$beta.exposure
  )
  
  index <- ss1$pval < inst_pval1 & ss2$pval < inst_pval2
  dat <- dat[index,]

  model1 <- '
  y1 ~ biv*x1
  y2 ~ biv*x2
  '
  mod1 <- sem(model1, data=dat)

  model2 <- '
  y1 ~ biv_1*x1
  y2 ~ biv_2*x2
  '
  mod2 <- sem(model2, data=dat)
  
  return(list(model1=mod1, model2=mod2))
}
```

Example 1 - the causal effects are the same

```{r}
set.seed(123456)
res <- sim1(nsnp=100, hsq1=0.4, hsq2=0.4, rg=0, nid1=20000, nid2=20000, maf1=0.4, maf2=0.5, biv1=0.2, biv2=0.2, inst_pval1=1, inst_pval2=1)
(s1 <- summary(res$model1, fit.measures=TRUE))
(s2 <- summary(res$model2, fit.measures=TRUE))
print(c(mod1=s1$FIT['aic'], mod2=s2$FIT['aic']))
```

According to https://www.scribbr.com/statistics/akaike-information-criterion/ "If a model is more than 2 AIC units lower than another, then it is considered significantly better than that model."

```{r}
s1$FIT['aic'] - s2$FIT['aic']
```

Example 2 - Now try with different causal effects

```{r}
set.seed(12345)
res <- sim1(200, 0.4, 0.4, 1, 20000, 20000, 0.4, 0.5, 0.2, -0.2, inst_pval1=1, inst_pval2=1)
(s1 <- summary(res$model1, fit.measures=TRUE))
(s2 <- summary(res$model2, fit.measures=TRUE))
print(c(mod1=s1$FIT['aic'], mod2=s2$FIT['aic']))
```

```{r}
s1$FIT['aic'] - s2$FIT['aic']
```

Clearly finds a big difference

---

## Real examples

```{r}
simple_multi_pop_sem <- function(exp1, exp2, out1, out2, inst_pval)
{
  exp <- mv_extract_exposures(c(exp1, exp2))
  dx <- inner_join(
    subset(exp, id.exposure == exp1),
    subset(exp, id.exposure == exp2),
    by="SNP"
  ) %>%
    dplyr::select(SNP=SNP, x1=beta.exposure.x, x2=beta.exposure.y, p1=pval.exposure.x, p2=pval.exposure.y) %>%
    dplyr::filter(p1 < inst_pval & p2 < inst_pval)

  plot1 <- dx %>%
    ggplot(., aes(x=x1, y=x2)) +
    geom_point()
  
  rgx <- cor(dx$x1, dx$x2)
  
  y <- extract_outcome_data(dx$SNP, c(out1, out2))
  dy <- inner_join(
    subset(y, id.outcome == out1),
    subset(y, id.outcome == out2),
    by="SNP"
  ) %>%
    dplyr::select(SNP=SNP, y1=beta.outcome.x, y2=beta.outcome.y)
  
  dat <- inner_join(dx, dy, by="SNP")

  model1 <- '
  y1 ~ biv*x2
  y2 ~ biv*x2
  '
  mod1 <- sem(model1, data=dat)

  model2 <- '
  y1 ~ biv_1*x2
  y2 ~ biv_2*x2
  '
  mod2 <- sem(model2, data=dat)

  s1 <- summary(mod1, fit.measures=TRUE)
  s2 <- summary(mod2, fit.measures=TRUE)

  dat2 <- harmonise_data(exp, y)
  
  mrres <- mr(dat2, method="mr_ivw") %>%
    dplyr::select(exposure, outcome, nsnp, b, se, pval)
  
  aic_diff <- s1$FIT['aic'] - s2$FIT['aic']
  
  return(list(dat=dat, plot1=plot1, rgx=rgx, mod1=s1, mod2=s2, mrres=mrres, aic_diff=aic_diff))
}
```

BMI and CHD

```{r}
bmi_chd <- simple_multi_pop_sem("bbj-a-1", "ukb-b-19953", "bbj-a-109", "ieu-a-7", 0.0005)
bmi_chd$mod1$PE
bmi_chd$mod2$PE
bmi_chd$aic_diff
```


Smoking and CHD

```{r}
smoking_chd <- simple_multi_pop_sem("bbj-a-78", "ukb-d-20116_0", "bbj-a-109", "ieu-a-7", 0.0005)
smoking_chd$mod1$PE
smoking_chd$mod2$PE
smoking_chd$aic_diff
```


---

Questions:

1. Power and false discovery rates to detect differences in causal effects in two populations by
  - Different sample sizes
  - Different allele frequencies
  - Different correlations between X1 and X2
  - Different heritabilities of X1 and X2
  - Different values of biv1 and biv2

2. Introduce winner's curse. What happens if pop1 has more power than pop2, and you only use SNPs that are significant at 5e-8


